### ISSUE TITLE
[no-undef] Exported interface is marked as unused

### ISSUE THREAD

**Issue by timdp** [typescript-eslint/typescript-eslint#131]
**What code were you trying to parse?**

https://github.com/timdp/standard-typescript-interfaces

**What did you expect to happen?**

I'm trying to use [Standard with TypeScript](https://standardjs.com/index.html#typescript).

The compiler allows me to export an interface from `types.ts` and import it into `index.ts`, but the linter and formatter complain.

My goal is actually to create a shared `types` package in a monorepo, but I've simplified it to a plain old module.

I'm well aware that interfaces don't exist in JavaScript, but given that `tsc` allows me to do this, ESLint should understand it as well, right?

**What actually happened?**

```
$ standard '*.ts'
standard: Use JavaScript Standard Style (https://standardjs.com)
  .../index.ts:1:10: 'Foo' is defined but never used.
  .../types.ts:11:10: 'Foo' is not defined.
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.1.0` |
| `TypeScript`                | `3.2.4` |
| `ESLint`                    | `5.4.0` |
| `node`                      | `10.15.0` |
| ~~`npm`~~ `yarn`            | `1.13.0` |

**Comment by j-f1** [comment]
Make sure you’ve enabled the `@typescript-eslint/no-unused-vars` rule from the `@typescript-eslint` plugin.

**Comment by timdp** [comment]
That would avoid the unused-import warning but the type _is_ actually used. Note the second error.

**Comment by j-f1** [comment]
Does installing the latest `@typescript-eslint/eslint-plugin` help?

**Comment by timdp** [comment]
I just updated the code to do that. Same error though.

**Comment by bradzacher** [comment]
Could you please chuck a succinct repl case into the OP?

It's a bit hard for us to reproduce and fix the issue if we have to checkout your entire repo to investigate.

**Comment by armano2** [comment]
^ 
> https://github.com/timdp/standard-typescript-interfaces

**Comment by bradzacher** [comment]
I'm assume `standard` just uses the recommended config?
This is what I mean by creating a succinct repl - the level of indirection with a tool makes it a bit harder to assess. Have to switch between a repo and its files..

The `no-unused-vars` error is probably due to the regressions from v1.1.0?

**Comment by bradzacher** [comment]
Reviewing the source for `no-undef`, it looks to me like a bug with the scope extensions that we're doing - the interface is not added to the global variable scope.

**Comment by timdp** [comment]
Cool, cool. Still need the REPL demo then?

**Comment by armano2** [comment]
@bradzacher to be precise, non of "type" nodes are not added to scope,
`type Foo`, `interface`, `typeParameters`, `typeArguments` and so on.

**Comment by mubaidr** [comment]
Any update?

**Comment by bradzacher** [comment]
I would suggest not using no-undef on typescript code.
The typescript compiler should catch all of the cases itself, so you shouldn't need the rule.

We'll look into a fix, though it will not be soon.
Scope analysis is a difficult thing when you add in the multiple and shared scopes that typescript adds.

**Comment by mubaidr** [comment]
^ Yes, makes sense. 👍

**Comment by meikidd** [comment]
Same issue here. Below is my code, it is allowed by typescript, but the linter deny it.

```
type User = {
  id: number;
  name: string;
};

export default User;
```

**Comment by mockdeep** [comment]
We're running into this same issue. And, unfortunately, TypeScript does not catch the undefined globals because in this case we're dealing with JQuery. In the type defs for JQuery, it is declared as a global, so TypeScript ignores references to it. For various reasons, we're trying to move to locally importing JQuery in our files, so the `no-undef` rule is pretty valuable to making sure we don't miss anything.

The workaround for the moment, was to move the type it was complaining about to being a global type def in a `.d.ts` file.

**Comment by juanmait** [comment]
I have the same issue and is also present in class constructors when trying to initialize the class properties directly in the constructor porps

```typescript
import { SomeInterface } from './some.module'

export class Hello {
  constructor(private readonly someProp: SomeInterface) {}

  getHello(): string {
    return this.someProp.getHello()
  }
}
```

In this code snippet I have two errors:

- In the `import` statement: `'SomeInterface' is defined but never used`.
- In the class constructor: `'someProp' is defined but never used`.

**Comment by bradzacher** [comment]
Merging into #1856
